### DISCUSSION TITLE
Programmatic workflow versionning

### DISCUSSION THREAD

**Discussion by ItsDyze** [elsa-workflows/elsa-core#2785] [Q&A]
Is it possible to get a list of rules that programmatic workflow follows in order to continue in their subsequent versions?
As far as I can tell it seems to remember the step it was in, but I can't tell how

**Comment by sfmskywalker** [comment]
Sure, here's how workflow execution works:

1. Given a workflow and optionally a starting activity ID;
2. Add the activity (the specified one, or the first one that does not have a parent) to a stack.
3. For as long as there are activities on the stack, pop them one by one;
4. For each activity, call its Execute method.
5. Before and after each activity execution, store the current activity ID in the workflow instance's `CurrentActivity` property. Also store the current stack of enqueued activities in the `ScheduledActivities` property.

Rule 5 is what ensures the engine knows how to resume workflow execution in case the server suddenly gets killed and needs to reboot & resume.

Assuming that no server accidents happen, all Elsa needs is the next activity to execute, and then it figures out what activity to execute next thanks to the workflow blueprint, which consists of a list of all activities on the workflow and their connections.
